### PR TITLE
[5.1] SEF: Implementing index.php behavior

### PR DESCRIPTION
**Pull request by Hackwar**
### Summary of Changes
Joomla has been improving its SEO performance constantly and one issue which is still open is the behavior of `index.php` in URLs. In some rare cases, the URL might end with `index.php`, even though that is not needed anymore, resulting in duplicate content, since the same URL without `index.php` gets the same result. A common case is the URL for the homepage, which often enough contains that reference. Another case is sites which switch from non-rewrite to rewrite the URLs.

This PR introduces one new setting in the SEF system plugin. This setting enforces removing the index.php from the URL with a redirect when the URL starts or ends with index.php. This setting only takes effect if URL rewriting is enabled. 

This PR depends on #42692.

I'd like to thank [ithelps Digital](https://www.ithelps-digital.com/) for sponsoring this feature.

### Testing Instructions
1. Apply the PR
2. Enable the option in the SEF system plugin.
3. Open a URL with /index.php/ at the beginning and see that the site is redirected to the version without.
6. Open a URL with /index.php at the end and see that the site is redirected to the version without.

### Link to documentations
Please select:
- [X] Documentation link for docs.joomla.org: https://docs.joomla.org/Search_Engine_Friendly_URLs
- [ ] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [ ] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Comment by brianteeman** [comment]
1. Instead of adding another option which a large % of users will never use why not simply do this in the htaccess.
2. the description is confusing "Remove unneeded index.php from URLs" because it sounds the same as the existing option which removes index.php
3. If accepted dont you need to update the installation and the configuration.php.dist
4. What happened to the requirement for all new "features" to be accompanied with documentation

**Comment by Hackwar** [comment]
To the other points: see other PR

To 2.: I will reword that.

File: administrator/language/en-GB/plg_system_sef.ini [hunk 1]
```diff
@@ -5,5 +5,8 @@
 
 PLG_SEF_DOMAIN_DESCRIPTION="If your site can be accessed through more than one domain enter the preferred (sometimes referred to as canonical) domain here. <br><strong>Note:</strong> https://example.com and https://www.example.com are different domains."
 PLG_SEF_DOMAIN_LABEL="Site Domain"
+PLG_SEF_INDEXPHP_DESCRIPTION="This option removes superfluous index.php in the URL and will only work when 'Use URL Rewriting' is enabled in Global Configuration."
+PLG_SEF_INDEXPHP_LABEL="Remove unneeded index.php from URLs"
+PLG_SEF_INDEXPHP_REDIRECT_LABEL="Redirect unnecessary index.php URLs"
```
Comment (Quy) [inline administrator/language/en-GB/plg_system_sef.ini]:
```suggestion
PLG_SEF_INDEXPHP_LABEL="Remove index.php from URLs"
PLG_SEF_INDEXPHP_REDIRECT_LABEL="Redirect index.php URLs"
```

**Comment by brianteeman** [comment]
I'm misunderstanding something

Why would you have the first option enabled but **not** have the second option enabled

**Comment by Hackwar** [comment]
The first option modifies how URLs are created. The second one modifies how URLs are parsed. Since people are doing awful things in routing, we can't be sure that their changes don't break this feature, this the 2 options can be enabled separately.

**Comment by brianteeman** [comment]
Honestly I cant see the value for core of adding 4 extra options (this PR and the other one) to address issues in other peoples code. Thats a maintainability nightmare and a usability issue for all those people that wonder what these options do etc. If, as you say, core does not create these urls then it should be for the extension dev to fix their own problem. And if they wont the site owner can easily create an htaccess rule.

**Comment by ot2sen** [comment]
Brian got quite a few valid points here. Why?
The instruction for test even tells why not. "Unfortunately for testing, core Joomla does not produce such URLs anymore, so we need to do some trickery to trigger the problem." Then why is it a core issue?

Wouldn´t it better go into a checking feature for the official extension JED Checker, for it to teach third-party devs how not to do awful things in routing? IE, do this and you won´t get approved for fame and showcase. Fix it and get in.
https://extensions.joomla.org/extension/jedchecker/

**Comment by brianteeman** [comment]
Adding more options is never a solution. Those same users will still complain as they dont know about the options (unless they ask). If its such a real problem (I have no idea) then make the changes without options

**Comment by brianteeman** [comment]
Confusing with the existing option

image

File: plugins/system/sef/src/Extension/Sef.php [hunk 2]
```diff
@@ -188,6 +245,51 @@ function ($match) use ($base, $protocols) {
...
+     * @param   Uri     &$uri     Uri object.
+     *
+     * @return  void
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    public function enforceRemovedIndexphp(&$router, &$uri)
+    {
+        // We only want to redirect on GET requests
+        if ($_SERVER['REQUEST_METHOD'] != 'GET') {
```
Comment (wilsonge) [inline plugins/system/sef/src/Extension/Sef.php]:
```suggestion
        if ($this->app->getInput()->getMethod() != 'GET') {
```

If i remember right this is the same thing but properly using the Input abstraction

Comment (Hackwar) [reply plugins/system/sef/src/Extension/Sef.php]: I'm going to trust you on this one.

**Comment by Hackwar** [comment]
I changed the feature and description above.

**Comment by SniperSister** [comment]
I have tested this item :white_check_mark: successfully on 4e471b01f9f282ebd85afe86e69b44014a1e5112<br /><br />Works as described!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42704">issues.joomla.org/tracker/joomla-cms/42704</a>.</sub>

**Comment by viocassel** [comment]
I have tested this item :white_check_mark: successfully on 4e471b01f9f282ebd85afe86e69b44014a1e5112<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42704">issues.joomla.org/tracker/joomla-cms/42704</a>.</sub>

**Comment by richard67** [comment]
Setting RTC as it has 2 good tests, but setting also the RMDG (release manager decision queue) label because this PR is subject of discussion among maintainers.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42704">issues.joomla.org/tracker/joomla-cms/42704</a>.</sub>

File: plugins/system/sef/src/Extension/Sef.php [hunk 3]
```diff
@@ -23,8 +27,57 @@
...
+    {
+        if (
+            is_a($event->getRouter(), SiteRouter::class)
+            && $this->app->get('sef_rewrite')
+        ) {
+            if ($this->params->get('indexphp')) {
+                // Enforce removing index.php with a redirect
+                $event->getRouter()->attachParseRule([$this, 'removeIndexphp'], SiteRouter::PROCESS_BEFORE);
+            }
+        }
```
Comment (Quy) [inline plugins/system/sef/src/Extension/Sef.php]:
```suggestion
            is_a($event->getRouter(), SiteRouter::class)
            && $this->app->get('sef_rewrite')
            && $this->params->get('indexphp')
        ) {
            // Enforce removing index.php with a redirect
            $event->getRouter()->attachParseRule([$this, 'removeIndexphp'], SiteRouter::PROCESS_BEFORE);
        }
```

File: plugins/system/sef/sef.xml [hunk 4]
```diff
@@ -30,6 +30,19 @@
...
+
+				<field
+					name="indexphp"
+					type="radio"
+					label="PLG_SEF_INDEXPHP_LABEL"
+					description="PLG_SEF_INDEXPHP_DESCRIPTION"
+					layout="joomla.form.field.radio.switcher"
+					default="0"
+					filter="boolean"
+				>
```
Comment (Quy) [inline plugins/system/sef/sef.xml:42]:
```suggestion
					>
```

File: plugins/system/sef/src/Extension/Sef.php [hunk 5]
```diff
@@ -188,6 +241,38 @@ function ($match) use ($base, $protocols) {
...
+     * @param   Uri     &$uri     Uri object.
+     *
+     * @return  void
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    public function removeIndexphp(&$router, &$uri)
+    {
+        // We only want to redirect on GET requests
+        if ($this->app->getInput()->getMethod() != 'GET') {
```
Comment (HLeithner) [inline plugins/system/sef/src/Extension/Sef.php]:
```suggestion
        if ($this->app->getInput()->getMethod() !== 'GET') {
```

File: plugins/system/sef/src/Extension/Sef.php [hunk 6]
```diff
@@ -188,6 +241,38 @@ function ($match) use ($base, $protocols) {
...
+            return;
+        }
+
+        $origUri = Uri::getInstance();
+
+        if (substr($origUri->getPath(), -9) == 'index.php') {
```
Comment (HLeithner) [inline plugins/system/sef/src/Extension/Sef.php]:
```suggestion
        if (substr($origUri->getPath(), -9) === 'index.php') {
```

File: plugins/system/sef/src/Extension/Sef.php [hunk 7]
```diff
@@ -188,6 +241,38 @@ function ($match) use ($base, $protocols) {
...
+            // Remove trailing index.php
+            $origUri->setPath(substr($origUri->getPath(), 0, -9));
+            $this->app->redirect($origUri->toString());
+        }
+
+        if (substr($origUri->getPath(), \strlen(Uri::base(true)), 11) == '/index.php/') {
```
Comment (HLeithner) [inline plugins/system/sef/src/Extension/Sef.php]:
```suggestion
        if (substr($origUri->getPath(), \strlen(Uri::base(true)), 11) === '/index.php/') {
```

**Review by HLeithner — CHANGES_REQUESTED** [review]

**Comment by richard67** [comment]
Code style fixes and strict comparison should be done. After that, I think it just needs a review but not new human tests.

**Comment by richard67** [comment]
Back to pending as there have been made changes.
@SniperSister @viocassel Could you test again? Thanks in advance.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42704">issues.joomla.org/tracker/joomla-cms/42704</a>.</sub>

**Comment by viocassel** [comment]
I have tested this item :white_check_mark: successfully on c35a3a0252205465fc9a14bd30a9b61c8d41e58e<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42704">issues.joomla.org/tracker/joomla-cms/42704</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42704">issues.joomla.org/tracker/joomla-cms/42704</a>.</sub>

**Comment by Hackwar** [comment]
Link to documentation has been added.

**Pull request merged**